Nav Unification - implement conditionals for static fallback nav data

mreishus opened the pull request:
This improves the fallback data response to more closely resemble what is returned by the API endpoint. Note however that we aren't trying to guarantee a 1:1 between the dynamic API items and these static fallback. These fallback are very much that - fallback! We just need to aim for "good enough". To this end I've decided to avoid adding too many conditionals, interogating state or relying on API requests to decided what should/shouldn't be shown.

The aim is to make the fallback acceptable enough so that on the few occasions the user does see it, they don't notice a huge difference. 

Closes #45910

## Demo

📺 https://d.pr/v/ph2Fpx/cyvkiVr3sC - for an example in action take a look at the screencapture I recorded. Slow down to 0.5x to see the transition between fallback and the response from the API.



## Changes proposed in this Pull Request

* (Already existing) Nav Unification implements a left sidebar menu that's rendered with data pulled from an AJAX request to wpcom.
* (Already existing) That sidebar is rendered from static data in calypso if the menu items ajax request fails.
* (In this PR) We attempt to add/remove some items from the fallback data depending on other information we can find in state. In general however we opt for static data and avoid over complicating things.

## Feedback Requested

**Note:** The following feedback has been updated by @getdave in response to changes in this PR:

* Does this make sense to do?
  * ~If data could not be loaded from AJAX, we are making decisions with other data loaded from different AJAX.  It seems like a niche failure case we're addressing (one ajax request fails while other ajax requests work).~ @getdave - Great point. After considering, yes I believe it does but _only in limited circumstances_. I've now dramatically cut back on where we're making API requests to determine what is/isn't shown in the fallback. We need to strike a happy medium here. In general we should aim for completely static data and this is what I have documented in the code comments.
  * ~Data isn't always available.  For example, I found testimonials and portfolio in `state.siteSettings.items`, but that isn't always filled out.  It's requested by `<QuerySiteSettings>` only on some pages.  It seems like a waste to add that request though, as this is only rarely needed.~ @getdave - I agree. Some data isn't available (eg: `state.siteSettings` until you navigate to a particular portion of the admin. Therefore we should avoid relying on such data and just accept we don't need to be complete 1:1 with the actual menu API response. Otherwise we're going to be rewriting the Menu API! 

* Note: This isn't complete.
  * ~`shouldShowTestimonials`, `shouldShowPortfolio`: Found these in `state.siteSettings`.~ Defaulted these to `true` and removed any attempt to determine dynamically.
  * ~`shouldShowLinks`: Couldn't find a case when links don't show.  Does anyone know them?~ Defaulted these to `true` and removed any attempt to determine dynamically.
  * ~`shouldShowWooCommerce`, `shouldShowApperanceHeaderAndBackground`, `shouldShowAdControl`, `shouldShowAMP`, `shouldShowThemeOptions`: Wanted to get some feedback before spending time on these.  I did default ad control and AMP to off, though, since it seems like that's the default case for these.~ I have defaulted most of these to `false` with the expectation of WooCommerce where I am checking whether the Plugin is active. I think this is important enough to be worth the trade off of an additional API request.

* Note: The PR might be easier to follow if you look at the individual commits.

## Testing instructions

* Have at least one blog with testimonials turned on. To turn on testimonials, switch to the twentytwenty theme, then in regular calypso, visit Sidebar -> Manage -> Settings -> Main Area "Writing" Tab -> Turn testimonials on.
* Check out this branch.
* Open Dev tools and under `Application` delete the IndexDB to clear it out.
* Visit calypso with Nav Unification active.
* Observe the menu items on the left. You should see a set of fallback menu items load and then be replaced by the items from the menu API response. 
* If you're not sure you are seeing the fallback items, you should try editing some of the titles of the items in `client/my-sites/sidebar-unified/fallback-data.js` and then repeating the test procedure (being sure to clear IndexDB between each run else you'll get the cache!).
* Also try the test procedure above with an eCommerce site active. WooCommerce should show up in the fallback menu data if it is active.

#### Alternative - force use fallback data

If you wish to for the fallback data for testing purposes you can manually edit `client/my-sites/sidebar-unified/use-site-menu-items.js` as follows:

```diff
+      return buildFallbackResponse( fallbackDataOverides );
-       return menuItems ?? buildFallbackResponse( fallbackDataOverides );
 }; 
```

mreishus committed:
fallback-data: Allow "shouldXYZ" variables to be false

The file builds up a menu like so:

	const fallbackResponse = [
		{
			icon: 'dashicons-admin-home',
			slug: 'home',
			title: translate( 'My Home' ),
			type: 'menu-item',
			url: `/home/${ siteDomain }`,
		},
		{
			icon: 'dashicons-chart-bar',
			slug: 'stats',
			title: translate( 'Stats' ),
			type: 'menu-item',
			url: `/stats/day/${ siteDomain }`,
		},
		...( shouldShowLinks && [
                    {
                        links,
                    },
                    {
                        related stuff here,
                    }
                ]),

The problem is, if "shouldShowLinks" was false or null, it would
generate invalid expressions and crash.  "...(false && [1, 2, 3])" would reduce
to "...(false)", and the spread iterator would attempt to iterate over
false, which would cause an error.

- Changed ...(shouldXYZ && []) to ...(shouldXYZ ? [] : [])
- Ran prettier

No functional changes.

mreishus committed:
fallback-data: Fix spelling errors

mreishus committed:
fallback-data: buildFallbackResponse() takes arguments

- No functional changes.
- Allow variables like shouldShowLinks to be passed in.

mreishus committed:
useSiteMenuItems fallback: Testimonals and Portfolio support are checked from site settings

mreishus committed:
use-site-menu-items fallback: Default AMP and AdControl to false

I haven't been able to find the exact conditions for these, but I think
in a lot of cases they're not applicable.

obenland commented on client/my-sites/sidebar-unified/fallback-data.js:
```diff
@@ -297,31 +304,35 @@ export default function buildFallbackResponse( { siteDomain = '' } = {} ) {
+								title: translate( 'Background' ),
+								type: 'submenu-item',
+								url: `https://${ siteDomain }/wp-admin/customize.php?return=%2Fwp-admin%2F&autofocus%5Bcontrol%5D=background_image`,
+							},
```
Themes can add support for one and not for the other, so these might have to be split

mreishus replied:
Did a little research, I found:
* On wpcom the backend checks `current_theme_supports( 'custom-header' )` and `( current_theme_supports( 'custom-background' )` separately.
* For an example of it, try using theme "Dara"
* I got theme-related information in calypso to load once, but I'm not sure how.  I think the selector would be something like this:

```js
	import { getActiveTheme, getCanonicalTheme } from 'calypso/state/themes/selectors';
	const theme = useSelector( ( state ) => {
		const currentThemeId = getActiveTheme( state, selectedSiteId );
		return {
			currentThemeId,
			currentTheme: getCanonicalTheme( state, selectedSiteId, currentThemeId ),
		};
	} );
```

The only part I was able to verify was `currentThemeId`, I wasn't able to see what `getCanonicalTheme` returned because I didn't know how I got the information loaded.

getdave replied:
We probably need to actively `dispatch` an action to fetch the Theme data before the store will be populated. Let me try that.

getdave replied:
I've removed the need for these tests. I feel it's getting overly complicated for the ROI.

obenland commented on client/my-sites/sidebar-unified/use-site-menu-items.js:
```diff
@@ -32,12 +33,31 @@ const useSiteMenuItems = () => {
+		shouldShowAdControl,
+		shouldShowAMP,
+		shouldShowThemeOptions,
+	};
```
It feels a little odd to me that we have this component decide which elements to show. Any reason `buildFallbackResponse()` couldn't figure that out itself?

mreishus replied:
We require hooks to look inside of global state + `buildFallbackResponse()` is called conditionally + hooks can't be called conditionally.

mreishus committed:
Put appearance/theme options behind edit_theme_options capability

- Also removed shouldShowThemeOptions as it didn't correspond to
  anything I could find in wpcom

getdave commented:
> shouldShowWooCommerce

Could we check if the Plugin is active? Or is there a Woo specific store that is added we can check for to determine whether or not to show?

getdave committed:
Add top level menu item for WooCommerce if active

getdave committed:
Programically check whether Woo is active before showing

getdave commented on client/my-sites/sidebar-unified/use-site-menu-items.js:
```diff
@@ -32,12 +38,42 @@ const useSiteMenuItems = () => {
-			siteDomain,
-		} )
+
+	const shouldShowLinks = true;
```
Still need to figure this one out.

getdave replied:
Defaulted to `true`.

getdave commented on client/my-sites/sidebar-unified/use-site-menu-items.js:
```diff
@@ -32,12 +38,42 @@ const useSiteMenuItems = () => {
+	const shouldShowThemes = useSelector( ( state ) =>
+		canCurrentUser( state, selectedSiteId, 'edit_theme_options' )
+	);
+	const shouldShowApperanceHeaderAndBackground = true;
```
Need to work on this one.

getdave replied:
Let's not worry about these. We're getting overly complicated for what is in essence a fallback.

getdave commented:
@cpapazoglou Any ideas how I can get this to use the correct icon for Jetpack?

<img width="487" alt="Screen Shot 2021-01-06 at 11 30 03" src="https://user-images.githubusercontent.com/444434/103763994-96871280-5012-11eb-8a2f-dce2cf647975.png">

getdave committed:
Split out header and background options.

getdave committed:
Rename Purchase to Upgrades to match menu data API response

getdave committed:
Allow all items to use defaults in buildFallbackResponse

Also removes unneeded comments

getdave committed:
Document purpose of fallback data

cpapazoglou commented:
> @cpapazoglou Any ideas how I can get this to use the correct icon for Jetpack?
> 
> <img alt="Screen Shot 2021-01-06 at 11 30 03" width="487" src="https://user-images.githubusercontent.com/444434/103763994-96871280-5012-11eb-8a2f-dce2cf647975.png">

Have you tried picking the icon as it is served from the endpoint? `data:image/svg+xml,%3Csvg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 32 32" %3E%3Cpath fill="%23a0a5aa" d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z"%3E%3C/path%3E%3Cpolygon fill="%23fff" points="15,19 7,19 15,3 "%3E%3C/polygon%3E%3Cpolygon fill="%23fff" points="17,29 17,13 25,13 "%3E%3C/polygon%3E%3C/svg%3E`

getdave committed:
Remove checks for testimonials and portfolios and improve docs

getdave committed:
Use correct JP icon

Commit:
Removes unused loadash dependency

Commit:
Adds upgrades subitems

Commit:
Removes jetpack scan

Commit:
Removes theme options submenu

cpapazoglou reviewed (approved):
Thanks for picking up and wrangling this @getdave!

It works pretty well! 

- I removed `import { get } from 'lodash';` from `client/my-sites/sidebar-unified/use-site-menu-items.js`, I hope you don't mind.
- Added Upgrades submenus.
- Removed Jetpack Scan which was leading to 404 (and shouldn't appear in Simple sites)
- Removed Theme Options which isn't standard and was leading to `/wp-admin`

We can merge this now! 🚢

getdave commented:
I will merge in the morning UTC.
